### PR TITLE
Fixes for #46 and #31

### PR DESCRIPTION
**Pull request by TSSlade**
I had introduced some problems in the __main__.py module via erroneous indentation of the revised handle_load method. This one fixes those. (Although not, presumably, any other errors I may have introduced.)

### CODE REVIEW

**Comment by bthayer2365** [comment]
Hey @TSSlade thanks for working on so much! I'll start off by reviewing this PR, then I'll move on and respond to your comments on other issues.

This PR contains all the commits of the previous two pull requests, (#56 and #57) so I'm going to rename this one close those.

If you want to work on other issues, and have me review them, you'll need to get them merged you'll have to base your changes off of my master branch instead of your own previous work.

Here's how to do that:
```
git remote add ben git@github.com:bthayer2365/git-gud.git
git fetch ben
git checkout ben/master
git checkout -b next-issue
```
I'll use github's review feature to comment on your work

fixes #46 
fixes #31

**Comment by bthayer2365** [comment]
Also, I see you're a first-time contributor and I'd like to welcome you to open source and Git Gud!

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -25,7 +25,7 @@ I am committed to that goal and more than happy to help to you learn to code and
 ```
 git clone https://github.com/bthayer2365/git-gud
 cd git-gud
-pip install -e . # Installs in edit mode so we can change files without needing to reinstall
+pip install -e . # NB: Do not overlook the `.` after the -e option! Installs in edit mode so we can change files without needing to reinstall
```
Comment (bthayer2365) [inline CONTRIBUTING.md]: Great comment. I can see how someone could miss that. Can you add this to README.md too?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -115,14 +115,14 @@ In this case, the parser knows that the previous commit is the parent and that t
 In the case of branches, the first commit on a branch will have a parent that is not the previous commit. 
 Instead, you'll need to include the colon and specify the parent commit. 
 Likewise, if a branch head or a tag points to a commit, then we must include the parenthesis, otherwise, we can leave it out. 
-The most complete example of a spec file is `gitgud/levels/merging/_octopus/test.spec`. 
+The most complete example of a spec file is `gitgud/extras/merging/_octopus/test.spec`.
```
Comment (bthayer2365) [inline CONTRIBUTING.md]: Should be `gitgud/levels/extras/_octopus/test.spec`

File: gitgud/__main__.py [hunk 3]
```diff
@@ -21,7 +21,7 @@ class InitializationError(Exception):
 
 class GitGud:
     def __init__(self):
-        self.file_operator = get_operator()  # Only gets operator if in a valid gitgud repo
+        self.file_operator = get_operator()  # Only gets operator if in a valid git
```
Comment (bthayer2365) [inline gitgud/__main__.py]: You said you'd change this back to the original text in your previous PR, but I don't see those changes, can you push them?

File: gitgud/__main__.py [hunk 4]
```diff
@@ -42,7 +42,7 @@ def __init__(self):
         instructions_parser = self.subparsers.add_parser('instructions', help='Show the instructions for the current level', description='Show the instructions for the current level')
         reset_parser = self.subparsers.add_parser('reset', help='Reset the current level', description='Reset the current level')
         test_parser = self.subparsers.add_parser('test', help='Test to see if you\'ve successfully completed the current level', description='Test to see if you\'ve successfully completed the current level')
-        progress_parser = self.subparsers.add_parser('progress', help='Continue to the next level', description='Continue to the next level')
+        progress_parser = self.subparsers.add_parser('progress', help='Continue to the next level', description='This command is now deprecated. To continue to the next level please use `git gud load` instead.')
```
Comment (bthayer2365) [inline gitgud/__main__.py]: We're in v0.1 pre-release right now, so no need to worry about deprecation, go ahead and delete this line. Same goes for anything else related to the progress command

File: gitgud/__main__.py [hunk 5]
```diff
@@ -177,18 +177,19 @@ def handle_test(self, args):
     def handle_progress(self, args):
```
Comment (bthayer2365) [inline gitgud/__main__.py]: You can delete this whole function

**Comment by bthayer2365** [comment]
For changing the levels, you'll need to also add the `__init__.py` in the new `extras/` folder. You'll also need to go through and make sure everything is properly registered, so look in at `levels/__init__.py` and in the other `__init__.py` left in `merging/`

**Comment by bthayer2365** [comment]
Also, thanks for fixing all those typos

**Comment by bthayer2365** [comment]
Make those changes and then we should be ready for a merge!

**Comment by bthayer2365** [comment]
I'll make sure tabs are okay after all the changes are made and I test the code

**Comment by bthayer2365** [comment]
Also, a note for any other changes you make, no need to create a new pull request every time you make changes. Pull requests track the branch. For example, this one tracks TSSlade:master. Any commits you make to master will show up here, which also means you should be careful when making other changes so that they don't show up here. Other Pull Requests can be made on different branches, making it easy to keep solutions for different issues isolated from one another.

**Comment by TSSlade** [comment]
Thanks for the feedback! First time contributing, and prior GitHub work (mostly for my own work/school stuff) hasn't involved lots of branching, so I remain unclear about how the branches and PRs play together. I'd actually thought I'd needed a different PR from each branch, but then it looked after the fact as if many of the changes I'd thought were in only one branch were in both. (hmmmm...)
Will make the adjustments tomorrow and then update.

**Comment by bthayer2365** [comment]
You're right about one PR per branch, your braches all just had the same changes. My guess is you wrote your changes, then created a new branch forgetting that the changes from that branch would be carried with it. Git log should be useful. Looking forward to the updates!

**Comment by bthayer2365** [comment]
We're also going to want to update the readme to remove git gud progress and to add git gud load next

File: CONTRIBUTING.md (hunk 2 above)
Comment (TSSlade) [reply CONTRIBUTING.md]: No `merging` in the path? Alrighty.

**Comment by TSSlade** [comment]
Okay, I think I've sorted out the feedback you provided above.

For future reference on branching: having now brought `master` up to speed, this is where I should be pulling from ben/master, then branching from there?

**Comment by bthayer2365** [comment]
That's right, and then you can have all other issues on their own branches, all based off of ben/master. Eventually, once this PR is merged, you'll be able to branch off of your own master again since your changes will be part of my history also and you can have your master up to date with mine again.

In general, I like to keep all my changes off of master, that way, I can just run git pull and then branch off of master, but you can do the exact same workflow with your master as is by running

```
git checkout ben/master
git pull ben master
git checkout -b feature
```
The only difference is that you're checking out my master instead of yours all the time. You'll still need to run the initial git fetch as I explained before in order to create ben/master, but after that things are basically the same as if it were your master, and arguably more clear, but also a little more typing.

I'll review your changes tomorrow afternoon.

**Comment by bthayer2365** [comment]
Hey @TSSlade, thanks for the updates. Can you:

 - [ ] Change the spec file on line 118 in CONTRIBUTING.md to `gitgud/levels/extras/_octopus/test.spec` You deleted it outright
 - [ ] Change `__main__.py` to add back `# Only gets operator if in a valid gitgud repo`
 - [ ] Move `levels/merging/__init__.py` to `levels/extras/__init__.py`

Also, can you clarify what `self.human_name` in `gitgud/levels/util.py` is trying to fix? If I recall, that's part of another issue, but can't remember which. Can you let me know which one?

**Comment by TSSlade** [comment]
image
Near as I can tell the first two are already done - this is a screenshot of my current status, with no changes made since my last PR.
Will make the __init__.py change right now.

**Comment by TSSlade** [comment]
And btw, the human_name thing was an abortive initial step at tackling #49.  Never got past that initial step, and indeed had not realized it was caught up in the PRs to date.

**Comment by bthayer2365** [comment]
Okay, thanks for getting back to me on this. Looks like you've made those changes, but they're not showing up on github and I double-checked locally, and when I clone, they don't seem to be there either. Can you check to make sure you actually committed/pushed your changes and that they're on the right branch? They should all be on master.

Also, for your changes for #49, can you put revert them here and put them in a different PR? We're going to need to add names to all the levels when we do that. 

If you need help, you can separate your changes in two ways: one way is to the git way, where you cherry-pick and rebase and then force push, or you can just change things manually and write new commits. You can go back and do the ben/master thing we talked about and make your human_name changes there, which you'd also have to do if trying to do the cherry-picking/rebasing stuff too.

**Comment by bthayer2365** [comment]
Hey, @TSSlade are you still working on this?

**Comment by TSSlade** [comment]
Hey! Yes, sorry, I am. Just got swamped by work-work and lost a couple days.
Have confirmed that I'd been pushing to an (apparently) non-existent branch. (I mean, it existed at one point. But what I did with it is anyone's guess.) So I was staring down the barrel of needing to either figure out how to retrieve the changes from elsewhere or wrap my head around your cherry-picking / rebase suggestion, and I just didn't have the bandwidth to do either. I'll take a look on Thurs.

**Comment by bthayer2365** [comment]
Hey @TSSlade, if you're still stuck on the rebasing, I can help. 

Go ahead and copy your master branch using 
```
git branch feature-46
git branch feature-31
git branch feature-49
```

Open a pull request for each of these and we can move forward from there.

I'll give you instructions for how to get each of the changes isolated to each of those branches, or if you'd like, I can do it myself. I'll also be able to push to your master branch and get it synced up with my master again.

**Comment by TSSlade** [comment]
Yeah, I'm pretty hopelessly confused right now. I'll take the guidance on getting the branches isolated, since that seems like something I'll need to understand in the future.

I'm attempting to pull `git pull ben master`, but it's failing because of unresolved conflicts...but I feel like i keep making the same edits / conflict resolutions, and then can't seem to figure out where to push changes to and then pull request from.

Barring that, I am wondering whether the easiest thing at this point would be to just nuke my local version of this thing, pull afresh from `ben/master`, introduce the new branches from the fresh pull, and then just re-do those edits from scratch.

**Comment by bthayer2365** [comment]
If you'd like to try your luck rebasing, go ahead and run the commands I just said and I can help you through the process. This would be a good opportunity to practice using interactive rebasing (`git rebase -i`), but I understand if that's too complicated. It's tough at first, but after a while, it starts making sense.

If you'd like to start from scratch and copy your changes manually, go ahead and run these:
```
git checkout master
git checkout -b backup
git push -u origin backup
git branch -d master
git fetch ben
git checkout ben/master
git checkout -b master
git push -f -u origin master
git branch feature-31
git branch feature-46
git branch feature-49
```
You'll have a clean start from my master branch. All your work will be saved in the backup branch and you'll be able to copy all your work on the feature branches you just created.

After looking at your repository, it seems like you've erased the changes you've made, although I'm not sure how, since I don't see the changes here in this PR. (Maybe you forked my repo again?)

Also, for reference, the head of your commit branch is at ef0bf0f73937b36fe6fe25347ab6406b8ab590c1. To check it out and go back to your changes, you can run
```
git checkout -b newbranch ef0bf0f
```

**Comment by TSSlade** [comment]
I wound up just deleting my repo, pulling yours again, and re-implementing the changes. Since they were trivial, it wasn't hard. I just don't know whether I missed anything along the way.

I'll read up on what you'd suggested above as the alternative approach. I'm sure I'll screw things up in future contributions and thus have an opportunity to practice that skill. :-)

**Comment by bthayer2365** [comment]
Haha, no problem. I'll review all the PRs and let you know if anything is missing. It might be fun to copy the repo folder (since it still has the git objects) and see if you can use the rebase commands to actually get it all worked out. As far as PRs go, do whatever you'd like, I'll be waiting!

**Comment by bthayer2365** [comment]
Closing this since you've made updates on other PRs.

**Pull request closed without merging**